File: shared/src/main/java/xyz/klinker/messenger/shared/util/MessageListStylingHelper.java
package xyz.klinker.messenger.shared.util;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.DrawableRes;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import xyz.klinker.messenger.shared.R;
import xyz.klinker.messenger.shared.data.FeatureFlags;
import xyz.klinker.messenger.shared.data.MimeType;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.data.model.Message;

public class MessageListStylingHelper {

    private boolean roundMessages;
    private int eightDp;

    private int columnType = -1;
    private int columnTimestamp = -1;

    private boolean isJustSentMessage;
    private int currentType;
    private long currentTimestamp;
    private int lastType;
    private long lastTimestamp;
    private int nextType;
    private long nextTimestamp;

    public MessageListStylingHelper(Context context) {
        try {
            eightDp = DensityUtil.toDp(context, 8);
            roundMessages = Settings.get(context).rounderBubbles;
        } catch (NullPointerException e) {
            eightDp = 8;
            roundMessages = false;
        }
    }

    public MessageListStylingHelper calculateAdjacentItems(Cursor cursor, int currentPosition) {
        if (columnTimestamp == -1) {
            columnTimestamp = cursor.getColumnIndex(Message.Companion.getCOLUMN_TIMESTAMP());
        }
        if (columnType == -1) {
            columnType = cursor.getColumnIndex(Message.Companion.getCOLUMN_TYPE());
        }

        cursor.moveToPosition(currentPosition);
        currentType = cursor.getInt(columnType);
        currentTimestamp = cursor.getLong(columnTimestamp);

        if (currentPosition > 0) {
            cursor.moveToPosition(currentPosition - 1);
            lastType = cursor.getInt(columnType);
            lastTimestamp = cursor.getLong(columnTimestamp);
        } else {
            lastType = -1;
            lastTimestamp = -1;
        }

        if (currentPosition != cursor.getCount() - 1) {
            cursor.moveToPosition(currentPosition + 1);
            nextType = cursor.getInt(columnType);
            nextTimestamp = cursor.getLong(columnTimestamp);
            isJustSentMessage = false;
        } else {
            isJustSentMessage = currentType != Message.Companion.getTYPE_RECEIVED();
            nextType = -1;
            nextTimestamp = System.currentTimeMillis();
        }

        return this;
    }

    public MessageListStylingHelper setMargins(View itemView) {
        if (itemView.getLayoutParams() == null) {
            return this;
        }

        if (currentType != lastType) {
            ((RecyclerView.LayoutParams) itemView.getLayoutParams()).topMargin = eightDp;
        } else {
            ((RecyclerView.LayoutParams) itemView.getLayoutParams()).topMargin = 0;
        }

        if (!isJustSentMessage && (currentType != nextType || TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp))) {
            ((RecyclerView.LayoutParams) itemView.getLayoutParams()).bottomMargin = eightDp;
        } else {
            ((RecyclerView.LayoutParams) itemView.getLayoutParams()).bottomMargin = 0;
        }

        return this;
    }

    public MessageListStylingHelper setBackground(View messageHolder, String mimeType) {
        if (MimeType.INSTANCE.isExpandedMedia(mimeType) || currentType == Message.Companion.getTYPE_INFO() || messageHolder == null) {

        } else if (mimeType.contains("image") || mimeType.contains("video")) {
            messageHolder.setBackground(null);
        } else {
            int background;
            if (roundMessages) {
                background = roundBubbleBackground();
            } else {
                background = dialogSquareBackground();
            }

            messageHolder.setBackground(
                    messageHolder.getContext().getResources().getDrawable(background)
            );
        }

        return this;
    }

    public MessageListStylingHelper applyTimestampHeight(TextView timestamp, int timestampHeight) {
        if (TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp)) {
            timestamp.getLayoutParams().height = timestampHeight;
        } else {
            timestamp.getLayoutParams().height = 0;
        }

        return this;
    }

    @DrawableRes
    private int dialogSquareBackground() {
        if (currentType == lastType && !TimeUtils.shouldDisplayTimestamp(lastTimestamp, currentTimestamp)) {
            if (currentType == Message.Companion.getTYPE_RECEIVED()) {
                return R.drawable.message_received_group_background;
            } else {
                return R.drawable.message_sent_group_background;
            }
        } else {
            if (currentType == Message.Companion.getTYPE_RECEIVED()) {
                return R.drawable.message_received_background;
            } else {
                return R.drawable.message_sent_background;
            }
        }
    }

    @DrawableRes
    private int roundBubbleBackground() {
        boolean displayNextTimestamp = TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp);
        boolean displayLastTimestamp = TimeUtils.shouldDisplayTimestamp(lastTimestamp, currentTimestamp);

        if (currentType == lastType && currentType == nextType && !displayLastTimestamp && !displayNextTimestamp) {
            // both corners
            if (currentType == Message.Companion.getTYPE_RECEIVED()) {
                return R.drawable.message_circle_received_group_both_background;
            } else {
                return R.drawable.message_circle_sent_group_both_background;
            }
        } else if ((currentType == lastType && currentType != nextType && !displayLastTimestamp) ||
                (currentType == nextType && currentType == lastType && displayNextTimestamp)) {
            // top corner bubble
            if (currentType == Message.Companion.getTYPE_RECEIVED()) {
                return R.drawable.message_circle_received_group_top_background;
            } else {
                return R.drawable.message_circle_sent_group_top_background;
            }
        } else if ((currentType == nextType && currentType != lastType && !displayNextTimestamp) ||
                (currentType == nextType && currentType == lastType && displayLastTimestamp)) {
            // bottom corner bubble
            if (currentType == Message.Companion.getTYPE_RECEIVED()) {
                return R.drawable.message_circle_received_group_bottom_background;
            } else {
                return R.drawable.message_circle_sent_group_bottom_background;
            }
        } else {
            return R.drawable.message_circle_background;
        }
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/util/MessageListStylingHelper.kt
package xyz.klinker.messenger.shared.util

import android.content.Context
import android.database.Cursor
import android.view.View
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.recyclerview.widget.RecyclerView

import xyz.klinker.messenger.shared.R
import xyz.klinker.messenger.shared.data.MimeType
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.model.Message
import xyz.klinker.messenger.shared.data.pojo.BubbleTheme

class MessageListStylingHelper(context: Context?) {

    private val eightDp: Int = DensityUtil.toDp(context, 8)

    private var columnType = -1
    private var columnTimestamp = -1
    private var columnContact = -1

    private var currentType: Int = 0
    private var currentTimestamp: Long = 0
    private var currentFrom: String? = null

    private var lastType: Int = 0
    private var lastTimestamp: Long = 0
    private var lastFrom: String? = null

    private var nextType: Int = 0
    private var nextTimestamp: Long = 0
    private var nextFrom: String? = null

    var hideContact = false


    fun calculateAdjacentItems(cursor: Cursor, currentPosition: Int): MessageListStylingHelper {
        if (columnTimestamp == -1) {
            columnTimestamp = cursor.getColumnIndex(Message.COLUMN_TIMESTAMP)
        }

        if (columnType == -1) {
            columnType = cursor.getColumnIndex(Message.COLUMN_TYPE)
        }

        if (columnContact == -1) {
            columnContact = cursor.getColumnIndex(Message.COLUMN_FROM)
        }

        cursor.moveToPosition(currentPosition)
        currentType = cursor.getInt(columnType)
        currentTimestamp = cursor.getLong(columnTimestamp)
        currentFrom = cursor.getString(columnContact)

        if (currentPosition > 0) {
            cursor.moveToPosition(currentPosition - 1)
            lastType = cursor.getInt(columnType)
            lastTimestamp = cursor.getLong(columnTimestamp)
            lastFrom = cursor.getString(columnContact)
        } else {
            lastType = -1
            lastTimestamp = -1
            lastFrom = null
        }

        if (currentPosition != cursor.count - 1) {
            cursor.moveToPosition(currentPosition + 1)
            nextType = cursor.getInt(columnType)
            nextTimestamp = cursor.getLong(columnTimestamp)
            nextFrom = cursor.getString(columnContact)
        } else {
            nextType = -1
            nextTimestamp = TimeUtils.now
            nextFrom = null
        }

        return this
    }

    fun setMargins(itemView: View): MessageListStylingHelper {
        if (itemView.layoutParams == null) {
            return this
        }

        if (!sameType(currentType, lastType) || !sameContact(currentFrom, lastFrom)) {
            (itemView.layoutParams as RecyclerView.LayoutParams).topMargin = eightDp
        } else {
            (itemView.layoutParams as RecyclerView.LayoutParams).topMargin = 0
        }

        if (!sameType(currentType, nextType) || !sameContact(currentFrom, nextFrom) || TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp)) {
            (itemView.layoutParams as RecyclerView.LayoutParams).bottomMargin = eightDp
        } else {
            (itemView.layoutParams as RecyclerView.LayoutParams).bottomMargin = 0
        }

        return this
    }

    fun setBackground(messageHolder: View?, mimeType: String): MessageListStylingHelper {
        if (MimeType.isExpandedMedia(mimeType) || currentType == Message.TYPE_INFO || messageHolder == null) {

        } else if (mimeType.contains("image") || mimeType.contains("video")) {
            messageHolder.background = null
        } else {
            val background = when (Settings.bubbleTheme) {
                BubbleTheme.SQUARE -> fourTypeBackground(if (currentType == Message.TYPE_RECEIVED) squareReceived else squareSent)
                BubbleTheme.ROUNDED -> fourTypeBackground(if (currentType == Message.TYPE_RECEIVED) roundReceived else roundSent)
                BubbleTheme.CIRCLE -> fourTypeBackground(if (currentType == Message.TYPE_RECEIVED) circleReceived else circleSent)
            }

            messageHolder.background = messageHolder.context.resources.getDrawable(background)
        }

        return this
    }

    fun applyTimestampHeight(timestamp: TextView, timestampHeight: Int): MessageListStylingHelper {
        if (TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp)) {
            timestamp.layoutParams.height = timestampHeight
        } else {
            timestamp.layoutParams.height = 0
        }

        return this
    }

    @DrawableRes
    private fun fourTypeBackground(set: DrawableHolder): Int {
        val displayNextTimestamp = TimeUtils.shouldDisplayTimestamp(currentTimestamp, nextTimestamp)
        val displayLastTimestamp = TimeUtils.shouldDisplayTimestamp(lastTimestamp, currentTimestamp)

        return if (same(currentFrom, lastFrom, currentType, lastType) && same(currentFrom, nextFrom, currentType, nextType) && !displayLastTimestamp && !displayNextTimestamp) {
            // both corners
            hideContact = true
            set.groupBoth
        } else if ((same(currentFrom, lastFrom, currentType, lastType) && !same(currentFrom, nextFrom, currentType, nextType) && !displayLastTimestamp) ||
                (same(currentFrom, nextFrom, currentType, nextType) && same(currentFrom, lastFrom, currentType, lastType) && displayNextTimestamp && !displayLastTimestamp)) {
            // top corner bubble
            hideContact = false
            set.groupTop
        } else if ((same(currentFrom, nextFrom, currentType, nextType) && !same(currentFrom, lastFrom, currentType, lastType) && !displayNextTimestamp) ||
                (same(currentFrom, nextFrom, currentType, nextType) && same(currentFrom, lastFrom, currentType, lastType) && displayLastTimestamp && !displayNextTimestamp)) {
            // bottom corner bubble
            hideContact = true
            set.groupBottom
        } else {
            // normal bubble
            hideContact = false
            set.noGrouped
        }
    }

    private fun same(contactOne: String?, contactTwo: String?, typeOne: Int, typeTwo: Int) = sameContact(contactOne, contactTwo) && sameType(typeOne, typeTwo)
    private fun sameContact(one: String?, two: String?) = one == two
    private fun sameType(one: Int, two: Int): Boolean {
        return when {
            one == two -> true
            one == Message.TYPE_SENDING && (two == Message.TYPE_SENT || two == Message.TYPE_DELIVERED || two == Message.TYPE_ERROR) -> true
            two == Message.TYPE_SENDING && (one == Message.TYPE_SENT || one == Message.TYPE_DELIVERED || one == Message.TYPE_ERROR) -> true
            else -> false
        }
    }


    private class DrawableHolder(val groupBoth: Int, val groupTop: Int, val groupBottom: Int, val noGrouped: Int)

    companion object {
        private val roundReceived = DrawableHolder(
                R.drawable.message_round_received_group_both_background,
                R.drawable.message_round_received_group_top_background,
                R.drawable.message_round_received_group_bottom_background,
                R.drawable.message_round_received_background
        )

        private val roundSent = DrawableHolder(
                R.drawable.message_round_sent_group_both_background,
                R.drawable.message_round_sent_group_top_background,
                R.drawable.message_round_sent_group_bottom_background,
                R.drawable.message_round_sent_background
        )

        private val circleReceived = DrawableHolder(
                R.drawable.message_circle_received_group_both_background,
                R.drawable.message_circle_received_group_top_background,
                R.drawable.message_circle_received_group_bottom_background,
                R.drawable.message_circle_background
        )

        private val circleSent = DrawableHolder(
                R.drawable.message_circle_sent_group_both_background,
                R.drawable.message_circle_sent_group_top_background,
                R.drawable.message_circle_sent_group_bottom_background,
                R.drawable.message_circle_background
        )

        private val squareReceived = DrawableHolder(
                R.drawable.message_square_received_group_background,
                R.drawable.message_square_received_group_background,
                R.drawable.message_square_received_background,
                R.drawable.message_square_received_background
        )

        private val squareSent = DrawableHolder(
                R.drawable.message_square_sent_group_background,
                R.drawable.message_square_sent_group_background,
                R.drawable.message_square_sent_background,
                R.drawable.message_square_sent_background
        )
    }
}


